scalapb-runtime/src/main/scala/scalapb/descriptors/Descriptor.scala

```diff
@@ -77,6 +77,12 @@ class Descriptor private[descriptors] (
       FieldDescriptor.buildFieldDescriptor(fd, index, this)
     }.toVector
 
+  lazy val fieldsByNumberMap: Map[Int, FieldDescriptor] =
```

Comment:
I wonder if this helps performance for your use case. In scalapb-json4s we are using this method to search within map entry (which have only two fields: key and value), or primitive wrappers (1 value field). 

Can you do a benchmark with this method reverted?

If I am missing something and this is used in larger messages, following protobuf-java which has been extensively optimized, an alternative to the map here could be another vector (or array) sorted by field number, and this method can do a binary search. https://github.com/protocolbuffers/protobuf/blob/7679378c7f20fb814d75ecd375e68fc32ffea9ba/java/core/src/main/java/com/google/protobuf/Descriptors.java#L809-L812 , Assuming your benchmark suggest this doesn't matter for now we can defer this part of the change.

Reply:
This is the actually the change that solves the O(n^2) complexity for json parsing (I only added the byName change for completeness, I can revert it if you like). `findFieldByNumber` is called in `messageReads` for every field in the message

Reply:
Ack, my bad. let's add `private` sorted arrays of the fd's, and introduce a private `@tailrec` annotated binary search. Add a unit tests to `ScalaDescriptorSpec` leveraging existing messages to check the binary search for edge cases: empty message, message with 1 field like a primitive wrapper, and a large one like `TestAllTypes` from `unittest.proto`.

Reply:
Why do you think a sorted array is better than a map?

Reply:
`Array[]` takes the least amount of memory possible; I anticipte lookup time to not vary match given the small size. I know Google's code is heavily benchmarked, so I'd opt for the same optimization they did if we're at it.